Sprint 2/step 2

Nikba-Creative-Studio opened the pull request:
**Build: https://demo5.nikba-dev.com/**

**1. Обновление инфраструктуры приложения**
Установите в проект все необходимые пакеты, которые отвечают за работу Redux-хранилища приложения и dnd. 

**2. Подготовка хранилища**
- список всех полученных ингредиентов,
- список всех ингредиентов в текущем конструкторе бургера,
- объект текущего просматриваемого ингредиента,
- объект созданного заказа.

**3. Создание первых экшенов и редьюсеров**
- Получение списка ингредиентов от API. Используется в компоненте BurgerIngredients.
- Получение списка ингредиентов для конструктора бургера. Используется в компоненте BurgerConstructor.
- Добавление данных о просматриваемом в модальном окне IngredientDetails ингредиенте.
- Удаление данных о просматриваемом в модальном окне ингредиенте при закрытии модального окна.
- Получение и обновление номера заказа в модальном окне OrderDetails.

**4. Доработка интерфейса навигации по ингредиентам**
- По мере пользовательского скролла ингредиентов в компоненте BurgerIngredients выделяйте активным тот переключатель, заголовок которого в самом контейнере ближе всего к верхней левой границе контейнера компонента BurgerIngredients.
- Нажатие на переключатель реализованно

**5. Реализация перетаскивания ингредиентов**
С помощью библиотеки react-dnd реализуйте такую функциональность:

- Пользователь может добавить ингредиент из BurgerIngredients в компонент BurgerConstructor.
- При успешном перетаскивании у ингредиента в BurgerConstructor увеличивается счётчик. Перетаскивать ингредиент (не являющийся булкой) можно многократно.
- Пользователь может нажать на иконку удаления ингредиента в компоненте BurgerConstructor. Ингредиент удалится из BurgerConstructor, а счётчик количества ингредиентов в компоненте BurgerIngredients уменьшится на один.
- Если в BurgerConstructor добавлено несколько одинаковых ингредиентов — удаление одного ингредиента не влияет на остальные ингредиенты в BurgerConstructor с тем же _id.
- При успешном перетаскивании ингредиента с типом bun из BurgerIngredients в BurgerConstructor текущий элемент с типом bun должен удалиться и замениться на перетаскиваемый. Все эти действия также должны отразиться на счётчике количества ингредиентов в BurgerIngredients.

**6. Вложенная сортировка ингредиентов в BurgerConstructor**
- пользователь может изменить порядок ингредиентов в BurgerConstructor перетаскиванием конкретного элемента;
- при попытке «бросить» ингредиент за пределы BurgerConstructor ничего происходить не должно — ингредиент возвращается в исходное положение.

Nikba-Creative-Studio committed:
Update burger-constructor.js

Рекомендаций код ревювера:
Не забудьте, пожалуйста, что у булок в собранном бургере должны быть иконки замков, а не корзинок.

Nikba-Creative-Studio committed:
Подготовка хранилища

1. Обновление инфраструктуры приложения
- react-redux redux-thunk redux
2. Подготовка хранилища
- rootReducer
- Redux Devtools

Nikba-Creative-Studio committed:
BurgerIngredients - redux

1. Получение списка ингредиентов от API. Используется в компоненте BurgerIngredients.
2. Добавление данных о просматриваемом в модальном окне IngredientDetails ингредиенте.
3. Удаление данных о просматриваемом в модальном окне ингредиенте при закрытии модального окна.

Nikba-Creative-Studio committed:
Доработка интерфейса навигации по ингредиентам

В компоненте BurgerIngredients есть три переключателя: «Булки», «Соусы» и «Начинки».

1. По мере пользовательского скролла ингредиентов в компоненте BurgerIngredients выделяйте активным тот переключатель, заголовок которого в самом контейнере ближе всего к верхней левой границе контейнера компонента BurgerIngredients.
2. Нажатие на переключатель реализовано

Nikba-Creative-Studio committed:
BurgerIngredients: Реализация перетаскивания ингредиентов

1. react-dnd react-dnd-html5-backend
2. Разработан фунционал для перетаскивания ингредиентов
3. Активация модального окна с описанием ингредиента перенесена в компонент Ingredient

Nikba-Creative-Studio committed:
BurgerConstructor: Реализация перетаскивания ингредиентов

1. Перетаскивание ингредиентов в конструктор бургеров
2. Удаление ингредиентов из конструктора бургеров
3. Булки выделены в отдельный массив

Nikba-Creative-Studio committed:
OrderDetails - Перенесен на Redux

1. Список ид, ингредиентов из конструктора передаются через post на бэк
2. При успешной отправке отрывается модальное окно с ид заказа

Nikba-Creative-Studio committed:
Счётчик количества ингредиентов в компоненте BurgerIngredients

Nikba-Creative-Studio committed:
Вложенная сортировка ингредиентов в BurgerConstructor

- пользователь может изменить порядок ингредиентов в BurgerConstructor перетаскиванием конкретного элемента;
- при попытке «бросить» ингредиент за пределы BurgerConstructor ничего происходить не должно — ингредиент возвращается в исходное положение.

Nikba-Creative-Studio committed:
Отчистка конструктора

После удачного отправления заказа, данные из конструктора удаляются.

Nikba-Creative-Studio committed:
Update burger-constructor.js

gennady-bars commented on src/services/store.js:
```diff
@@ -0,0 +1,16 @@
+const enhancer = composeEnhancers(applyMiddleware(thunk));
+
+// Create Redux store
+export const store = createStore(rootReducer, enhancer);
```
Отлично, что `store` в отдельном файле

gennady-bars commented on src/services/reducers/order-details.js:
```diff
@@ -0,0 +1,27 @@
+
+export const orderDetailsReducer = (state = orderDetailsInitialState, action) => {
+    switch (action.type) {
+        case 'ORDER/POST_ORDER_SUCCESS':
```
`'ORDER/POST_ORDER_SUCCESS'`

это строка


Каждый тип экшена должен быть вынесен в константу в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const GET_INGREDIENTS = 'GET_INGREDIENTS'`


Это нужно исправить во всем проекте

gennady-bars commented on src/services/reducers/burger-constructor.js:
```diff
@@ -0,0 +1,57 @@
+
+            return {
+                ...state,
+                ingredients: update(state.ingredients, {
```
`update` является внешней функцией, поэтому ее нельзя использовать внутри редьюсера. Редьюсер должен быть чистой функцией. Вам нужно это делать в `actions`, а данные передавать уже в виде `payload` в редьюсер

gennady-bars commented on src/services/actions/burger-constructor.js:
```diff
@@ -0,0 +1,50 @@
+export const CLEAN_CONSTRUCTOR = 'CONSTRUCTOR/CLEAN_CONSTRUCTOR';
+
+export function setIngredients(item, uid) {
+    return function(dispatch) {
```
**Можно лучше**

`thunk`-функции создаются обычно для асинхронных запросов на сервер, а для обычных экшенов делают обычные `action creator` (функцию, которая возвращает объект с `type`). И главный показатель – это то, что если внутри используется только 1 раз `dispatch`, то уже можно сделать  `action creator`, а не `thunk`
Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)


Вот так выглядит экшен-криейтер этой функции

```js



export function setIngredients(item, uid) {
    return {
        type: SET_INGREDIENTS,
        item: {...item, uid: uid}
    }
}



```

gennady-bars commented on src/services/actions/burger-ingredients.js:
```diff
@@ -0,0 +1,68 @@
+export const DESELECT_INGREDIENT = 'BURGER/DESELECT_INGREDIENT';
+
+//API Url
+const API_URL = 'https://norma.nomoreparties.space/api/ingredients'
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и использовать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.

gennady-bars commented on src/services/actions/order-details.js:
```diff
@@ -0,0 +1,64 @@
+export const HIDE_MODAL = 'ORDER/HIDE_MODAL';
+
+//Post API Url
+const API_ORDER_URL = 'https://norma.nomoreparties.space/api/orders'
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и использовать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.

gennady-bars commented on src/services/actions/burger-ingredients.js:
```diff
@@ -0,0 +1,68 @@
+    }
+}
+
+export function fetchIngredientsFailure(error) {
```
Вот это отличные `action creator`ы

gennady-bars commented on src/services/actions/burger-ingredients.js:
```diff
@@ -0,0 +1,68 @@
+
+        fetch(API_URL)
+            .then(response => {
+                if (!response.ok) {
```
Надо было проверять ответ вот так, чтобы сразу `res.json()` было внутри

```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется этот код во всех запросах. Чтобы этого избежать, нужно сделать отдельный метод `checkResponse`, который будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

и теперь можно удалить кучу ненужно повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только `ссылка` на метод. Не нужно его вызывать. Он сам вызовется, так как в `then` нужно передавать именно `функцию`, а не `вызов` функции.

gennady-bars commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,93 +1,72 @@
+                <Ingredient item={constructorBunsData} type='top' isLocked={true} />
+                
+                <div className={styles.ingredients}>
+                    {constructorData.map((item, index) => ( <Ingredient  key={uuid()}  item={item} isLocked={false} id={item._id} index={index} moveIngredient={moveIngredient} /> ))}
```
` key={uuid()} ` нельзя генерировать уникальный `uuid` прямо в `key`. Вы тут постоянно новый `key` создаете, и реакт не знает теперь, как следить за перемещениями компонента внутри заказа. 


для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время сброса в зону `DnD`. Это можно делать с помощью библиотеки `uuid`

gennady-bars reviewed (commented):
Здравствуйте. Работа проделана огромная, но есть некоторые недочеты:

* Файл `App.test.js` нужно удалить, так как он не относится к проекту
* при переносе ингердиента снизу заказа на верх в консоли ошибка вылезает. Скрин http://joxi.ru/52apb6YTgVg5Dm
* Каждый тип экшена должен быть вынесен в константу в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const GET_INGREDIENTS = 'GET_INGREDIENTS'` . Это нужно исправить во всем проекте
* `update` является внешней функцией, поэтому ее нельзя использовать внутри редьюсера. Редьюсер должен быть чистой функцией. Вам нужно это делать в `actions`, а данные передавать уже в виде `payload` в редьюсер
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и использовать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.
* нужно сделать отдельный метод `checkResponse`, который будет проверять ответ от сервера
* для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время сброса в зону `DnD`. Это можно делать с помощью библиотеки `uuid` . В идеале нужно делать это в `action creator`е в `actions`



**Можно лучше**

* `thunk`-функции создаются обычно для асинхронных запросов на сервер, а для обычных экшенов делают обычные `action creator` (функцию, которая возвращает объект с `type`). И главный показатель – это то, что если внутри используется только 1 раз `dispatch`, то уже можно сделать  `action creator`, а не `thunk`
Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Nikba-Creative-Studio committed:
Рефакторинг

1. Файл App.test.js удален
2. при переносе ингредиента снизу заказа на верх вылезала ошибка - исправлено, было связано с атрибутом key и uuid
3. Рефакторинг экшенов, в switch используется название переменой экшона
4. фунция update вместе с библиотекой immutability-helper удаленно
5. Адрес на бэкэнд перенесен в отдельную константу baseUrl в /utils/config.js. Используется константа + эндпоит
6. Разработан отдельный метод checkResponse в /utils/helpers.js для проверки ответа от севера
7. В GurgerConstructor ингредиенты получают уникальны атрибут key который задается при збросе ингредиента в зону конструктора
8. thunk функции остались только для запросов в бэкэнду, в остальном использовал action creator

Nikba-Creative-Studio committed:
Ошибка выполнении запроса к API

Если при выполнении запроса к API в усилителе произошла ошибка, то содержимое хранилища соответствующего
элемента приводится к начальному состоянию.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
